File: db4o.net/Db4objects.Db4o/native/Db4objects.Db4o/Foundation/CRC32.cs
namespace Db4objects.Db4o.Foundation
{
    public class CRC32
    {
        private static uint[] crcTable;

        static CRC32()
        {
            BuildCRCTable();
        }

        private static void BuildCRCTable()
        {
            uint Crc32Polynomial = 0xEDB88320;
            uint i;
            uint j;
            uint crc;
            crcTable = new uint[256];
            for (i = 0; i <= 255; i++)
            {
                crc = i;
                for (j = 8; j > 0; j--)
                {
                    if ((crc & 1) == 1)
                    {
                        crc = ((crc) >> (1 & 0x1f)) ^ Crc32Polynomial;
                    }
                    else
                    {
                        crc = crc >> (1 & 0x1f);
                    }
                }
                crcTable[i] = crc;
            }
        }

        public static long CheckSum(byte[] buffer, int start, int count)
        {
            uint temp1;
            uint temp2;
            int i = start;
            uint crc = 0xFFFFFFFF;
            while (count-- != 0)
            {
                temp1 = (crc) >> (8 & 0x1f);
                temp2 = crcTable[(crc ^ buffer[i++]) & 0xFF];
                crc = temp1 ^ temp2;
            }
            return (long)~crc & 0xFFFFFFFFL;
        }
    }
}

File: db4oj/core/src/com/db4o/foundation/CRC32.java
package com.db4o.foundation;

/**
 * @sharpen.ignore
 */
public class CRC32
{
   private static int crcTable[];

   static {
      buildCRCTable();     
   }

   private static void buildCRCTable()
   {
      final int CRC32_POLYNOMIAL = 0xEDB88320;

      int i, j;
      int crc;

      crcTable = new int[256];

      for (i = 0; i <= 255; i++)
      {
         crc = i;
         for (j = 8; j > 0; j--)
            if ((crc & 1) == 1)
               crc = (crc >>> 1) ^ CRC32_POLYNOMIAL;
            else
               crc >>>= 1;
         crcTable[i] = crc;
      }
   }

   public static long checkSum(byte buffer[], int start, int count)
   {
      int temp1, temp2;
      int i = start;

      int crc = 0xFFFFFFFF;

      while (count-- != 0)
      {
         temp1 = crc >>> 8;
         temp2 = crcTable[(crc ^ buffer[i++]) & 0xFF];
         crc = temp1 ^ temp2;
      }

      return (long) ~crc & 0xFFFFFFFFL;
   }
}
